Use ArraySlice to slice a Vector

jdunkerley opened the pull request:
[ci no changelog needed]

### Pull Request Description

Use an `ArraySlice` to slice `Vector`.
Avoids memory copying for the slice function.

### Important Notes
| Test | Ref | New |
| --- | --- | --- |
| New Vector | 71.9 | 71.0 |
| Append Single | 26.0 | 27.7 |
| Append Large | 15.1 | 14.9 |
| Sum | 156.4 | 165.8 |
| Drop First 20 and Sum | 171.2 | 165.3 |
| Drop Last 20 and Sum | 170.7 | 163.0 |
| Filter | 76.9 | 76.9 |
| Filter With Index | 166.3 | 168.3 |
| Partition | 278.5 | 273.8 |
| Partition With Index | 392.0 | 393.7 |
| Each | 101.9 | 102.7 |

- Note: the performance of New and Append has got slower from previous tests.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

jdunkerley committed:
Fix benchmark.

jdunkerley committed:
First pass at ArraySlice.

jdunkerley committed:
Java Format.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso:
```diff
@@ -660,10 +660,10 @@ type Vector a
+                Vector.from_polyglot_array (self.slice_builtin slice_start slice_end)
+
+    ## PRIVATE
+    slice_builtin self start end = @Builtin_Method "Vector.slice_builtin"
```
Interesting test: 
```
v = [1, 2, 3]
Meta.type_of v . should_equal Meta.type_of (v.slice 1 1)
```

jdunkerley replied:
no `type_of` yet so did similar tests.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso:
```diff
@@ -660,10 +660,10 @@ type Vector a
-                arr = Array.new len
-                Array.copy self.to_array slice_start arr 0 len
-                Vector.from_polyglot_array arr
+                Vector.from_polyglot_array (self.slice_builtin slice_start slice_end)
```
`Vector.from_polyglot_array` wraps the slice into yet another object.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java:
```diff
@@ -0,0 +1,113 @@
+  private final long length;
+
+  public ArraySlice(Object storage, long start, long length) {
+    if (storage instanceof ArraySlice slice) {
```
This condition is unlikely to be `true` - the code in https://github.com/enso-org/enso/pull/3724/files#r976005007 seems to wrap `ArraySlice` into yet another `Vector`.

jdunkerley committed:
Making slice return a Vector.

jdunkerley committed:
Java format.

hubertp committed:
InteropLibrary parameter is added in specialized

Parameters of type InteropLibrary are currently correctly generated only
when annotated with `Builtin.Specialized`. Should eventually get rid of
this limitation and merge the logic into both variants.

jdunkerley committed:
Fix built in name.

jdunkerley committed:
Few slice tests.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java:
```diff
@@ -0,0 +1,113 @@
+      this.start = slice.start + start;
+      this.length = length;
+    } else {
+      if (CompilerDirectives.inInterpreter()) {
+        if (!InteropLibrary.getUncached().hasArrayElements(storage)) {
+          throw new IllegalStateException("Vector needs array-like delegate, but got: " + storage);
+        }
+      }
```
I think I don't understand why this check is performed only in the interpreter. Won't this affect semantics depending on optimization levels?

radeusgd replied:
Anyway, if left as-is I think it would be beneficial to add a short comment explaining why this is only checked in interpreter.

Unless there is some obvious reason that every Truffle developer should understand immediately, I think such an explanation will help future developers understand this decision.

JaroslavTulach replied:
We could do the check as `assert`. Asserts are ignored when Truffle does its PE compilation. We just need to make sure we run our tests with `-ea`.

radeusgd replied:
Why can't the check be there _always_?

jdunkerley committed:
end not length!

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java:
```diff
@@ -67,6 +63,26 @@ public final Object toArray() {
+  @Builtin.Specialize
+  @Builtin.WrapException(from = UnsupportedMessageException.class, to = PanicException.class)
+  public final Vector slice(long start, long end, InteropLibrary interop)
+      throws UnsupportedMessageException {
```
In what circumstances is `UnsupportedMessageException` thrown?

jdunkerley replied:
Only if the backing storage no longer can answer the `interop.getArraySize` call - shouldn't happen.

radeusgd commented on test/Benchmarks/src/Vector.enso:
```diff
@@ -1,4 +1,5 @@
 from Standard.Base import all
+import Standard.Base
```
Why is this needed?

jdunkerley replied:
As the Benchmark is called `Vector` means you can access the `Standard.Base.Vector` as `Base.Vector`.

radeusgd replied:
Ah, okay. I was confused as there weren't any other changes.

But I assume before `from Standard.Base import all` was introducing `Base` into scope and this changed in one of the recent PRs and the benchmark was just not tested since then. So now it makes sense, thanks for the explanation.

radeusgd commented on test/Tests/src/Data/Vector_Spec.enso:
```diff
@@ -212,6 +212,15 @@ spec = Test.group "Vectors" <|
+        vec.slice 0 3 . should_equal [1, 2, 3]
+        vec.slice 1 3 . should_equal [2, 3]
+        vec.slice 1 1 . should_equal []
+        vec.slice 0 100 . should_equal [1, 2, 3, 4, 5, 6]
```
```suggestion
        vec.slice 0 100 . should_equal vec
```

JaroslavTulach replied:
Radek, do you suggest to test for identity? Then the check has to use `Meta.is_same_object`. Otherwise `==` on _Arrays&Vectors_ always only compares content.

radeusgd replied:
No, I suggest the check on content, I just don't see a reason for writing `[1, 2, 3, 4, 5, 6]` twice if its exactly the content of `vec`. But I get your point that this could _falsely_ suggest a check on _identity_ so probably good to leave it as-is.

radeusgd reviewed (approved):
Looks good to me overall, just a few small comments and a question on implementation that would be good if can be answered before merge.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java:
```diff
@@ -0,0 +1,113 @@
+  public long getArraySize(@CachedLibrary(limit = "3") InteropLibrary interop)
+      throws UnsupportedMessageException {
+    long storageSize = interop.getArraySize(storage);
+    return Math.min(storageSize, end) - start;
```
This could get negative, right?

Creates slice over JavaScript array and then remove something from the underlying array. The `storageSize` gets smaller and substracting `start` may give us negative index.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java:
```diff
@@ -0,0 +1,113 @@
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.interop.*;
```
I'd replace the star import with individual classes imports in the name of clarity.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java:
```diff
@@ -0,0 +1,113 @@
+  private final long start;
+  private final long end;
+
+  public ArraySlice(Object storage, long start, long end) {
```
Who's supposed to call this constructor? Just `Vector.slice` from the same package? Then make it package private. Possibly also make the whole class package private.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java:
```diff
@@ -67,6 +63,26 @@ public final Object toArray() {
+  @Builtin.Method(name = "slice", description = "Returns a slice of this Vector.")
+  @Builtin.Specialize
+  @Builtin.WrapException(from = UnsupportedMessageException.class, to = PanicException.class)
+  public final Vector slice(long start, long end, InteropLibrary interop)
```
Is the `InteropLibrary` cached or uncached?

hubertp replied:
If you look at the generated node it is cached

JaroslavTulach reviewed (approved):
Not bad. If that works, it can go in. There is `VectorBenchmarks.java` - consider adding a `sliced` check in there.

jdunkerley committed:
PR comments.

jdunkerley committed:
Slice benchmark.

jdunkerley committed:
Fix new benchmark.

mergify[bot] committed:
Merge branch 'develop' into wip/jd/vector-slice-182880080
